### PR TITLE
[core] Enable/Disable AddTagsListener

### PR DESCRIPTION
**Pull request by guillaumedelre**
Add attribute cache_invalidation in ApiResource

| Q             | A
| ------------- | ---
| Bug fix?      | no 
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tickets       | 
| License       | MIT

My customer uses varnish cache to maximize performance, but for some cases, I need to disable http cache invalidation.

**Proposal**: Enable/Disable AddTagsListener at runtime

It could be done **globally** with something like that:

```php
<?php
// api/src/Entity/Book.php

namespace App\Entity;

use ApiPlatform\Core\Annotation\ApiResource;

/**
 * ...
 * @ApiResource(
 *     cacheInvalidation=false,
 * ...
 * )
 */
class Book
{
    // ...
}
```

It could be disabled **at operation level** with something like that:

```php
<?php
// api/src/Entity/Book.php

namespace App\Entity;

use ApiPlatform\Core\Annotation\ApiResource;

/**
 * ...
 * @ApiResource(
 *     itemOperations={
 *         "get"={
 *             "cache_invalidation"=false,
 *             ...
 *         },
 *     },
 * ...
 * )
 */
class Book
{
    // ...
}
```

### CODE REVIEW

**Comment by guillaumedelre** [comment]
This is my very first PR, I hope someone will helps me :smile:

**Comment by soyuka** [comment]
Related issues: https://github.com/api-platform/core/issues/2926 https://github.com/api-platform/core/issues/3168 ping @bastnic

File: src/Annotation/ApiResource.php [hunk 1]
```diff
@@ -175,6 +177,14 @@ final class ApiResource
      */
     private $cacheHeaders;
 
+    /**
+     * @see https://api-platform.com/docs/core/performance/#enabling-disabling-http-cache-invalidation
```
Comment (soyuka) [inline src/Annotation/ApiResource.php:181]: not needed

File: src/HttpCache/EventListener/AddTagsListener.php [hunk 2]
```diff
@@ -49,6 +57,8 @@ public function onKernelResponse(ResponseEvent $event): void
             !$request->isMethodCacheable()
             || !$response->isCacheable()
             || (!$attributes = RequestAttributesExtractor::extractAttributes($request))
+            || !$attributes['cache_invalidation']
+            || $this->isOperationAttributeDisabled($attributes, self::OPERATION_ATTRIBUTE_KEY)
```
Comment (soyuka) [inline src/HttpCache/EventListener/AddTagsListener.php:61]: I think that the default should be `true` to keep the currnet behavior?

File: src/Annotation/ApiResource.php (hunk 1 above)
Comment (guillaumedelre) [reply src/Annotation/ApiResource.php]: done

**Comment by antograssiot** [comment]
Shouldn't this be done in PurgeHttpCacheListener listener also to avoid computing tags that needs to be purged ?
Also maybe `cacheTagsInvalidation` or something similar would be more accurate ?

**Comment by guillaumedelre** [comment]
> Also maybe `cacheTagsInvalidation` or something similar would be more accurate ?
my next commit will change the naming for `cacheTagsInvalidation`.

> Shouldn't this be done in PurgeHttpCacheListener listener also to avoid computing tags that needs to be purged ?
Something like that could be acceptable in the PurgeHttpCacheListener ?
- inject the requestStack 
- use ToggleableOperationAttributeTrait
- use the private method `isCacheTagsInvalidationEnabled` in the binded events

```php
final class PurgeHttpCacheListener
{
    use ToggleableOperationAttributeTrait;

    private $purger;
    private $iriConverter;
    private $resourceClassResolver;
    private $propertyAccessor;
    private $requestStack;
    private $tags = [];
even if I don't like this coupling
    public function __construct(PurgerInterface $purger, IriConverterInterface $iriConverter, ResourceClassResolverInterface $resourceClassResolver, RequestStack $requestStack, PropertyAccessorInterface $propertyAccessor = null)
    {
        $this->purger = $purger;
        $this->iriConverter = $iriConverter;
        $this->resourceClassResolver = $resourceClassResolver;
        $this->requestStack = $requestStack;
        $this->propertyAccessor = $propertyAccessor ?? PropertyAccess::createPropertyAccessor();
    }

    private function isCacheTagsInvalidationEnabled(): bool
    {
        $request = $this->requestStack->getCurrentRequest();

        return (
            (!$attributes = RequestAttributesExtractor::extractAttributes($request))
            || !$attributes['cache_tags_invalidation']
            || $this->isOperationAttributeDisabled($attributes, 'cache_tags_invalidation', false)
        ) ? false : true;
    }

    /**
     * Collects tags from the previous and the current version of the updated entities to purge related documents.
     */
    public function preUpdate(PreUpdateEventArgs $eventArgs): void
    {
        if (!$this->isCacheTagsInvalidationEnabled()) {
            return;
        }
```

**Comment by guillaumedelre** [comment]
@soyuka @antograssiot : 
Or in the `ApiPlatform\Core\EventListener\WriteListener`, we enrich the `$attributes` array with a new key that **allow us to disable** the listener `PurgeHttpCacheListener` in the `ApiPlatform\Core\Bridge\Doctrine\Common\DataPersister`


`ApiPlatform\Core\EventListener\WriteListener`: 
```php
...
        if (
            $controllerResult instanceof Response
            || $request->isMethodSafe()
            || !($attributes = RequestAttributesExtractor::extractAttributes($request))
            || !$attributes['persist']
            || $this->isOperationAttributeDisabled($attributes, self::OPERATION_ATTRIBUTE_KEY)
        ) {
            return;
        }

        $attributes['purge_http_cache_listener_disabled'] = $attributes['cache_tags_invalidation'] || $this->isOperationAttributeDisabled($attributes, 'cache_tags_invalidation', false);

        if (!$this->dataPersister->supports($controllerResult, $attributes)) {
            return;
        }
...
```

`ApiPlatform\Core\Bridge\Doctrine\Common\DataPersister`:
```php
    private $managerRegistry;
    private $purgeHttpCacheToggler;

    public function __construct(ManagerRegistry $managerRegistry, PurgeHttpCacheToggler $purgeHttpCacheToggler)
    {
        $this->managerRegistry = $managerRegistry;
        $this->purgeHttpCacheToggler = $purgeHttpCacheToggler;
    }
...

    /**
     * {@inheritdoc}
     */
    public function persist($data, array $context = [])
    {
        if (!$manager = $this->getManager($data)) {
            return $data;
        }

        if ($context['purge_http_cache_listener_disabled']) {
            $this->purgeHttpCacheToggler->disablePurgeHttpCacheListener();
        }

        if (!$manager->contains($data) || $this->isDeferredExplicit($manager, $data)) {
            $manager->persist($data);
        }

        $manager->flush();
        $manager->refresh($data);

        if ($context['purge_http_cache_listener_disabled']) {
            $this->purgeHttpCacheToggler->enablePurgeHttpCacheListener();
        }

        return $data;
    }

    /**
     * {@inheritdoc}
     */
    public function remove($data, array $context = [])
    {
        if (!$manager = $this->getManager($data)) {
            return;
        }

        if ($context['purge_http_cache_listener_disabled']) {
            $this->purgeHttpCacheToggler->disablePurgeHttpCacheListener();
        }

        $manager->remove($data);
        $manager->flush();

        if ($context['purge_http_cache_listener_disabled']) {
            $this->purgeHttpCacheToggler->enablePurgeHttpCacheListener();
        }
    }
...
```

et le nouveau service `ApiPlatform\Core\Bridge\Doctrine\Common\PurgeHttpCacheToggler`:
```php
<?php

declare(strict_types=1);

namespace ApiPlatform\Core\Bridge\Doctrine\Common;

use ApiPlatform\Core\Bridge\Doctrine\EventListener\PurgeHttpCacheListener;
use Doctrine\Common\EventManager;

final class PurgeHttpCacheToggler
{
    private $eventManager;

    public function __construct(EventManager $eventManager)
    {
        $this->eventManager = $eventManager;
    }

    private function getListener(): ?PurgeHttpCacheListener
    {
        foreach ($this->eventManager->getListeners() as $event => $listeners) {
            foreach ($listeners as $key => $listener) {
                if (!$listener instanceof PurgeHttpCacheListener) {
                    continue;
                }
                return $listener;
            }
        }

        return null;
    }

    public function disablePurgeHttpCacheListener():void
    {
        $this->eventManager->removeEventListener(['onFlush', 'preUpdate', 'postFlush'], $this->getListener());
    }

    public function enablePurgeHttpCacheListener(): void
    {
        $this->eventManager->addEventListener(['onFlush', 'preUpdate', 'postFlush'], $this->getListener());
    }
}
```

**Comment by soyuka** [comment]
I'm not a fan of the second solution you proposed, I tend to prefer the first even though I don't like the coupling either. 
I think we need @dunglas input on this.

**Comment by guillaumedelre** [comment]
hello @dunglas could youo give me inputs on this feature ? thank you

**Comment by vasilvestre** [comment]
I would like to help here if possible ! @dunglas are you available for input ?

**Comment by dunglas** [comment]
I need to take a deeper look. In my opinion the cleanest way would be to skip generating tags related to resources having invalidation disabled. However we must measure if computing metadata of all traversed resource types doesn't introduce an overhead (metadata should be cached, but we must double check to be sure).

Any help on implementing this patch and measuring its impact (using Blackfire for instance) will be very appreciated!

**Pull request closed without merging**